nixos/modules/services/misc/matrix-dendrite.nix

```diff
@@ -0,0 +1,171 @@
+            '';
+          };
+          disable_federation = lib.mkOption {
+            type = lib.types.bool;
+            default = true;
+            description = ''
+              Disables federation. Dendrite will not be able to make
+              any outbound HTTP requeststo other servers and the federation
+              API will not be exposed.
+            '';
+          };
+        };
+        options.client_api = {
+          registration_disabled = lib.mkOption {
+            type = lib.types.bool;
+            default = true;
+            description = ''
+              Whether to disable user registration to the server
+              without the shared secret.
+            '';
+          };
+        };
+      };
+      default = { };
+      description = ''
+        Configuration for dendrite, see:
+        <link xlink:href="https://github.com/matrix-org/dendrite/blob/master/dendrite-config.yaml"/>
+        for available options with which to populate settings.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.services.matrix-dendrite = {
+      description = "Dendrite Matrix homeserver";
+      after = [
+        "network.target"
+      ];
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        Type = "notify";
+        DynamicUser = true;
+        StateDirectory = "matrix-dendrite";
+        WorkingDirectory = workingDir;
+        ExecStart = lib.strings.concatStringsSep " " ([
+          "${pkgs.matrix-dendrite}/bin/dendrite-monolith-server"
+          "--config ${configurationYaml}"
+        ] ++ lib.optionals (cfg.httpPort != null) [
+          "--http-bind-address :${builtins.toString cfg.httpPort}"
+        ] ++ lib.optionals (cfg.httpsPort != null) [
+          "--https-bind-address :${builtins.toString cfg.httpsPort}"
+        ] ++ lib.optionals (cfg.tlsCert != null && cfg.tlsKey != null) [
+          "--tls-cert ${cfg.tlsCert}"
+          "--tls-key ${cfg.tlsKey}"
+        ]);
+        ExecReload = "kill -HUP $MAINPID";
+        Restart = "on-failure";
+      };
+    };
+    systemd.services.matrix-dendrite-key-generator =
+      lib.mkIf (cfg.generatePrivateKey || cfg.generateTls ) {
+        description = "Dendrite Matrix homeserver - Key Generator";
+        wantedBy = [ "matrix-dendrite.service" ];
+        requiredBy = [ "matrix-dendrite.service" ];
+        before = [ "matrix-dendrite.service" ];
+        path = with pkgs; [ matrix-dendrite ];
+        serviceConfig = {
+          Type = "oneshot";
+          DynamicUser = true;
+          StateDirectory = "matrix-dendrite";
+          WorkingDirectory = workingDir;
+          RemainAfterExit = true;
+          ExecStart = lib.strings.concatStringsSep " " ([
+            "${pkgs.matrix-dendrite}/bin/generate-keys"
+          ] ++ lib.optionals cfg.generatePrivateKey [
+            "--private-key ${cfg.settings.global.private_key}"
+          ] ++ lib.optionals (cfg.generateTls && cfg.tlsCert && cfg.tlsKey) [
```

Comment:
Assuming https://github.com/NixOS/nixpkgs/pull/109561#discussion_r563351017 is implemented, you can just use `lib.optionals cfg.generateTls` here

Reply:
Done